### PR TITLE
test: added anvil-catalog filter search test (#4125)

### LINKED ISSUE #4125: Add "search all filters" test to AnVIL-Catalog
For each tab, add the following two tests

Test 1 should:
- Select an arbitrary filter
- Read an arbitrary filter option
- Exit the filter menu
- Search for that filter, and select it
- Ensure filter label appears and that filter appears checked in the filter menu
- Repeat for 3-4 filters

Test 2 should:
- Select some filters through the filter menu
- Deselect those filters through the search all filters bar

### PR DESCRIPTION
**Pull request by mrtopsyt**
Added "Search all filters" tests to AnVIL-Catalog

### Ticket
Closes #4125 .

### Reviewers
@MillenniumFalconMechanic .

### Changes
- Added filter related constants to `anvilcatalog-tabs.ts`
- Added a test to
  - Select an arbitrary filter
  - Read the first filter option
  - Exit the filter menu
  - Search for that filter, and select it
  - Ensure filter label appears and that filter appears checked in the filter menu
  - Repeat for 3 filters
- Added another test to
  - Select an arbitrary filter
  - Check the first filter option
  - Exit the filter menu
  - Search for that filter, and deselect it
  - Ensure that the filter label appears unchecked
  - Repeat for 3 filters

### CODE REVIEW

File: explorer/e2e/testFunctions.ts [hunk 1]
```diff
@@ -431,6 +443,18 @@ export async function testFilterCounts(
   return true;
 }
 
+/**
+ * Get a locator for a named filter tag
+ * @param page - a Playwright page object
+ * @param filterTagName - the name of the filter tag to search for
+ * @returns - a locator for the named filter tag
+ */
```
Comment (MillenniumFalconMechanic) [inline explorer/e2e/testFunctions.ts:560]: Thank you (here and throughout)!

File: explorer/e2e/testFunctions.ts [hunk 2]
```diff
@@ -301,27 +301,39 @@ export async function testFilterPresence(
...
+export const getNamedFilterOptionLocator = (
   page: Page,
   filterOptionName: string
 ): Locator => {
-  return page
-    .getByRole("button")
-    .filter({ has: page.getByRole("checkbox"), hasText: filterOptionName });
+  return page.getByRole("button").filter({
+    has: page.getByRole("checkbox"),
+    hasText: RegExp(`^${escapeRegExp(filterOptionName)}\\s*\\d+\\s*`),
```
Comment (MillenniumFalconMechanic) [inline explorer/e2e/testFunctions.ts:407]: Would you be able to add comments describing the why behind the regex pattern here?

File: explorer/e2e/testFunctions.ts [hunk 3]
```diff
@@ -301,27 +301,39 @@ export async function testFilterPresence(
...
  */
-export const getFirstFilterButtonLocator = (page: Page): Locator => {
+export const getFirstFilterOptionLocator = (page: Page): Locator => {
   return page
     .getByRole("button")
     .filter({ has: page.getByRole("checkbox") })
     .first();
 };
 
+export const getFilterOptionName = async (
```
Comment (MillenniumFalconMechanic) [inline explorer/e2e/testFunctions.ts:423]: Could you add TSDocs for this function, and add a bit of detail around the [split/find](https://github.com/DataBiosphere/data-browser/pull/4135/files#diff-faefc5cfc7d7c7283d514ad866678143558aefea3eccb636b8bb7a1ae4abbae7R332-R333) functionality?

File: explorer/e2e/testFunctions.ts [hunk 4]
```diff
@@ -431,6 +443,18 @@ export async function testFilterCounts(
...
 
+/**
+ * Get a locator for a named filter tag
+ * @param page - a Playwright page object
+ * @param filterTagName - the name of the filter tag to search for
+ * @returns - a locator for the named filter tag
+ */
+const getFilterTagLocator = (page: Page, filterTagName: string): Locator => {
+  return page
+    .locator("#sidebar-positioner")
```
Comment (MillenniumFalconMechanic) [inline explorer/e2e/testFunctions.ts]: Thoughts on adding a const for `#sidebar-positioner`?

File: explorer/e2e/testFunctions.ts [hunk 5]
```diff
@@ -515,11 +535,101 @@ export async function testClearAll(
...
   }
 }
+
+/**
+ * Escape a string so it can safely be used in a regexp
+ * @param string - the string to escape
+ * @returns - A string that has all Regexp special characters escaped
+ */
+function escapeRegExp(string: string): string {
+  return string.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
```
Comment (MillenniumFalconMechanic) [inline explorer/e2e/testFunctions.ts:661]: Nice! Would you be able to add a comment describing the "why" behind the pattern used here?

**Review by MillenniumFalconMechanic — CHANGES_REQUESTED** [review]
Looks great, thanks Jonah! I've added a couple of requests for comments but LGTM other than that!

File: explorer/e2e/testFunctions.ts (hunk 4 above)
Comment (mrtopsyt) [reply explorer/e2e/testFunctions.ts]: I was able to remove one use of it (it's not recommended by Playwright as it's implementation dependent), and I changed the other to a constant!

File: explorer/e2e/testFunctions.ts (hunk 5 above)
Comment (mrtopsyt) [reply explorer/e2e/testFunctions.ts:661]: Will do! (it's copy-pasted from Findable lol)

**Review by MillenniumFalconMechanic — APPROVED** [review]
LGTM, thanks Jonah!

**Pull request merged**